### ISSUE TITLE
Setting the visibility of the restored method is in the wrong place

### ISSUE THREAD

**Issue by floehopper** [freerange/mocha#256]
The following code in `ClassMethod#restore_original_method` should be moved inside the `if @original_method && @original_method.owner == stubbee.__metaclass__` statement:

``` ruby
if @original_visibility
  Module.instance_method(@original_visibility).bind(stubbee.__metaclass__).call(method)
end
```

Otherwise it's vaguely conceivable that we might try to set the visibility of a method which does not exist. Or at least even if this isn't the case, this is a more logical place for the code to be.

**Comment by chrisroos** [comment]
This was fixed in 40a6eb2df2bace2283f3b13eff48b9589dcae8e4. Closing.

**Comment by chrisroos** [comment]
My mistake. Commit 40a6eb2df2bace2283f3b13eff48b9589dcae8e4 hasn't been merged to master yet so I'm reopening this issue.

@floehopper: Is there more work to do in the only-set-visibility-of-restored-method-if-it-has-been-newly-defined branch before it can be merged to master?

**Comment by floehopper** [comment]
@chrisroos:

> Is there more work to do in the only-set-visibility-of-restored-method-if-it-has-been-newly-defined branch before it can be merged to master?

Yes. There are failing tests on that branch in Ruby v1.8.7 & v1.9.3 - https://travis-ci.org/freerange/mocha/builds/169285775.

**Comment by chrisroos** [comment]
OK cool, @floehopper.

**Comment by nitishr** [comment]
Wondering if this is still relevant.

**Comment by floehopper** [comment]
@nitishr I think it _might_ still be relevant. I think the point I was trying to make is summed up in the commit note of the un-merged commit 40a6eb2:

> There's no point in considering setting the visibility of the restored method if we haven't actually restored (i.e. re-defined) it.

For the current state of the code in `master`, I think this would mean that `ClassMethod#restore_original_method` would look something like this:

```ruby
    def restore_original_method
      return if use_prepended_module_for_stub_method?
      if stub_method_overwrites_original_method?
        original_method_owner.send(:define_method, method_name, original_method_body)
        retain_original_visibility(original_method_owner)
      end
    end
```

Does that makes sense to you?

**Comment by nitishr** [comment]
Yes. I remember being puzzled by the application failing a number of tests when I made that change since it seemed logical that you'd only restore visibility if you've actually changed (and therefore restored) the method.
